classes/Divisions.php

```diff
@@ -0,0 +1,179 @@
+         * votes so we need to generate some text using the title of the division
+         */
+
+        switch ( strtolower($vote) ) {
+            case 'yes':
+            case 'aye':
+                $description = $this->constructYesNoVoteDescription('yes', $division_title, $yes_text);
+                break;
+            case 'no':
+                $description = $this->constructYesNoVoteDescription('no', $division_title, $no_text);
+                break;
+            case 'absent':
+                $description = ' was absent for a vote on ' . $division_title;
+                break;
+            case 'both':
+                $description = ' abstained on a vote on ' . $division_title;
+                break;
+            default:
+                $description = $division_title;
+        }
+
+        return $description;
+    }
+
+    private function divisionsByPolicy($q) {
+        $policies = array();
+
+        $row_count = $q->rows();
+        for ($n = 0; $n < $row_count; $n++) {
+            $policy_id = $q->field($n, 'policy_id');
+
+            if ( !array_key_exists($policy_id, $policies) ) {
+                $policies[$policy_id] = array(
+                    'policy_id' => $policy_id,
+                    'weak_count' => 0,
+                    'divisions' => array()
+                );
+            }
+            $division = array();
+
+            $direction = $q->field($n, 'direction');
+            if ( strpos( $direction, 'strong') !== FALSE ) {
+                $division['strong'] = TRUE;
+            } else {
+                $division['strong'] = FALSE;
+            }
+
+            $vote = $q->field($n, 'vote');
+            $yes_text = $q->field($n, 'yes_text');
+            $no_text = $q->field($n, 'no_text');
+            $division_title = $q->field($n, 'division_title');
+
+            $division['text'] = $this->constructVoteDescription($vote, $yes_text, $no_text, $division_title);
+            $division['division_id'] = $q->field($n, 'division_id');
+            $division['date'] = $q->field($n, 'division_date');
+            $division['vote'] = $vote;
+            $division['gid'] = fix_gid_from_db($q->field($n, 'gid'));
+            $division['url'] = $this->divisionUrlFromGid($q->field($n, 'gid'));
+
+            if ( !$division['strong'] ) {
+                $policies[$policy_id]['weak_count']++;
+            }
+
+            $policies[$policy_id]['divisions'][] = $division;
+        };
+
+        return $policies;
+    }
+
+    private function divisionUrlFromGid($gid) {
+        $q = $this->db->query("SELECT gid FROM hansard WHERE epobject_id = ( SELECT subsection_id FROM hansard WHERE gid = :gid )", array( ':gid' => $gid ));
+        $parent_gid = $q->field(0, 'gid');
+        if ( !$parent_gid ) {
+            return '';
+        }
+        $page_type = preg_replace('#.*/(.*)/.*#', '\1', $parent_gid);
+        $parent_gid = fix_gid_from_db($parent_gid);
+        $url = new \URL($page_type);
```

Comment:
I think a (slightly) better way of getting the URL would be to fetch the `major` in the query too, and then use `$hansardmajors[major]['page']`.